### ISSUE TITLE
Given very bad advice for a correction.

### ISSUE THREAD

**Issue by donatj** [scrutinizer-ci/php-analyzer#628]
I received the message:

```
The expression $options of type array is implicitly converted to a boolean; are you sure this is intended? If so, consider using ! empty($expr) instead to make it clear that you intend to check for an array without elements.
```

this is bad advice because empty checks for any sort of false-y value including $expr being unset.  To explicitly check if an array has values one should call `count($expr) !== 0`

> Here, you can explain why the found issue was invalid, and what result
> you would have expected otherwise.
>
> Please also check that this issue has not already been reported.

see https://scrutinizer-ci.com/g/capdig/capstone/inspections/9316395a-09aa-4d4e-b22b-61c62d4a55ea/issues/files/src/Router/WebRuleRouter.php?status=new&orderField=path&order=asc&issueId=78098756-544d-4542-a59d-63ea837d5bae

**Comment by jaspernbrouwer** [comment]
+1

see https://scrutinizer-ci.com/g/simple-es/event-sourcing/issues/master/files/src/Event/Store/InMemoryEventStore.php?orderField=path&order=asc&issueId=7ed0e5f0-6af3-4ae7-a8c7-4a1c8fb24462

I agree this is bad advise, but for a different reason:

`empty($foo)` has the same behavior as `!isset($foo) || !$foo`. In other words: The real difference between `empty()` and `!` is that `empty()` also checks if its argument actually exists (is defined/declared) and returns `false` if not, while `!` will issue a notice:

```php
!$foo;               // Undefined variable: foo on line 1
empty($foo);         // No errors

$foo = [];
!$foo['bar'];        // PHP error:  Undefined index: bar on line 5
empty($foo['bar']);  // No errors
```

Consider the following code (which is effectively what's going on in my link):

```php
$foo = [];
// ...

if (!$foo) {
    // ...
}
```

`!$envelopes` has no possible way of issuing a notice, because `$envelopes` is most definitely defined.

IMHO, and many other PHP developers with me, `empty()` should not be used in these cases.

PS: I also agree with @donatj that if you _truly_ want to express the intention of checking whether an array has no items (is empty), `count($foo) !== 0` would be the correct advice.

 `empty($foo)` would "make it clear that you intend to check for an array without elements" __in the English language__, but we're dealing with __PHP__ here. I think that's quite vital in this case.
